# Data Preparation

## 1. Import Pandas Library and Read in the csv-File

In [1]:
import pandas as pd

In [2]:
charging_data = pd.read_csv("charging_sessions.csv")

## 2. Data Dimensionality

Evaluate the dataset concerning dimensionality, data types and overall structure at first glance. 

In [3]:
charging_data.info(show_counts=True)
charging_data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

## 3. Transform Time into Datetime

Every datetime entry is originally in UTC. To analyze the data in local time, we need to convert these values to the timezone of each charging site.

In [4]:
date_columns = ["connectionTime", "disconnectTime", "doneChargingTime"]

for data in date_columns:
    charging_data[data] = pd.to_datetime(charging_data[data])

def convert_to_local(row): 
    tz = row["timezone"]
    for col in date_columns:
        ts = row[col]
        if pd.isna(ts):
            continue 
        row[col] = ts.tz_convert(tz)
    return row 

charging_data = charging_data.apply(convert_to_local, axis=1)

charging_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype                              
---  ------            --------------  -----                              
 0   Unnamed: 0        66450 non-null  int64                              
 1   id                66450 non-null  object                             
 2   connectionTime    66450 non-null  datetime64[ns, America/Los_Angeles]
 3   disconnectTime    66450 non-null  datetime64[ns, America/Los_Angeles]
 4   doneChargingTime  62362 non-null  datetime64[ns, America/Los_Angeles]
 5   kWhDelivered      66450 non-null  float64                            
 6   sessionID         66450 non-null  object                             
 7   siteID            66450 non-null  int64                              
 8   spaceID           66450 non-null  object                             
 9   stationID         66450 non-null  object                     

## 4. Duplicate Values

Check if "id" values are unique

In [5]:
duplicate_ids = charging_data["id"].duplicated(keep="first")
duplicate_ids = len(duplicate_ids[duplicate_ids == True])
duplicate_ids

1413

Hence, there are 1413 duplicate IDs. Further inspection of the duplicates is required before determining how to handle them.

In [ ]:
duplicated_ids = charging_data[charging_data["id"].duplicated()]
duplicated_ids = duplicated_ids["id"].unique()

subset_cols = charging_data.columns.difference(["Unnamed: 0"])

duplicates = charging_data[charging_data.duplicated(subset=subset_cols)]
len(duplicates)

1413

This shows that the entries with duplicate IDs also contain duplicate values for all other columns. Hence, the duplicates can be deleted without loss of information.

In [7]:
charging_data = charging_data.drop_duplicates(subset=subset_cols, keep="first")
charging_data.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Index: 65037 entries, 0 to 65036
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype                              
---  ------            --------------  -----                              
 0   Unnamed: 0        65037 non-null  int64                              
 1   id                65037 non-null  object                             
 2   connectionTime    65037 non-null  datetime64[ns, America/Los_Angeles]
 3   disconnectTime    65037 non-null  datetime64[ns, America/Los_Angeles]
 4   doneChargingTime  60950 non-null  datetime64[ns, America/Los_Angeles]
 5   kWhDelivered      65037 non-null  float64                            
 6   sessionID         65037 non-null  object                             
 7   siteID            65037 non-null  int64                              
 8   spaceID           65037 non-null  object                             
 9   stationID         65037 non-null  object                          

Check if "Unnamed: 0" values are unique

In [8]:
duplicate_unnamed = charging_data["Unnamed: 0"].unique()
print(len(duplicate_unnamed))

15292


The column "Unnamed: 0" ranges from 0 to 15292 and then repeats. Therefore we remove it completly, since its just another unnecessary index, that is even incorrect.

In [9]:
charging_data.drop(columns="Unnamed: 0", inplace=True)

This shows that the entries with duplicate IDs also contain duplicate values for all other columns. Hence, the duplicates can be deleted without loss of information. 

## 5. Invalid Data

Possible invalid scenarios are:

1. connectionTime is after disconnectTime
2. connectionTime is after doneChargingTime
3. The same stationID has overlapping sessions

1. Scenario

In [10]:
scenario1 = charging_data["connectionTime"] > charging_data["disconnectTime"]
scenario1.value_counts()

False    65037
Name: count, dtype: int64

No invalid data found.

2. Scenario

In [11]:
scenario2 = charging_data["connectionTime"] > charging_data["doneChargingTime"]
scenario2.value_counts()

False    65010
True        27
Name: count, dtype: int64

Invalid data found. It is physically not possible that the "doneChargingTime" is before the "connectionTime" so we delete these entires.

In [12]:
charging_data.drop(charging_data[charging_data["connectionTime"] > charging_data["doneChargingTime"]].index, inplace = True)

3. Scenario

In [13]:
charging_data["next_connection"] = charging_data.groupby("spaceID")["connectionTime"].shift(-1)
charging_data["next_disconnect"] = charging_data.groupby("spaceID")["disconnectTime"].shift(-1)

charging_data["overlap"] = charging_data["disconnectTime"] > charging_data["next_connection"]
charging_data["overlap"].value_counts()


overlap
False    64810
True       200
Name: count, dtype: int64

Invalid data found. It is also physically not possible that the same Space has two overlapping sessions so we delete these entries as well as the new created column "overlap".

In [14]:
charging_data.drop(charging_data[charging_data["overlap"]].index, inplace=True)

charging_data.drop(columns="overlap", inplace=True)

charging_data.head()

id            connectionTime  \
0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54-08:00   
1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50-08:00   
2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35-08:00   
3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58-08:00   
4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01-08:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 11:11:15-08:00 2020-01-02 09:31:35-08:00        25.016   
1 2020-01-02 14:38:21-08:00 2020-01-02 12:18:05-08:00        33.097   
2 2020-01-02 16:39:22-08:00 2020-01-02 08:35:06-08:00         6.521   
3 2020-01-02 08:38:39-08:00 2020-01-02 07:18:45-08:00         2.355   
4 2020-01-02 14:08:40-08:00 2020-01-02 10:17:30-08:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...   

            next_connection           next_disconnect  
0 2020-01-02 11:30:53-08:00 2020-01-02 15:18:39-08:00  
1 2020-01-02 16:35:27-08:00 2020-01-02 18:42:08-08:00  
2 2020-01-06 05:58:18-08:00 2020-01-06 13:52:37-08:00  
3 2020-01-02 08:48:03-08:00 2020-01-02 14:18:12-08:00  
4 2020-01-02 14:32:14-08:00 2020-01-02 19:36:55-08:00

## 6. Missing Values

Check for any missing values within the dataset

In [15]:
missing_values = charging_data.isnull().sum()
missing_values

id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4081
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              17175
userInputs          17175
next_connection       107
next_disconnect       107
dtype: int64

Quick analysis of the missing values:
- **"userID"** and **"userInputs"**: Contain missing values as expected, since users are not required to register
- **"doneChargingTime"**: Missing values are expected to be data irregularities and have to be dealt with in order to ensure a high data quality

Subsets of the data are created, to differentiate between regular and irregular (missing "doneChargingTime" value)

In [16]:
charging_data_regular_trips = charging_data[charging_data['doneChargingTime'].notnull()]
charging_data_wo_doneChargingTime = charging_data[charging_data['doneChargingTime'].isnull()]

print(f"Regular trips: {charging_data_regular_trips.shape[0]}")
print(f"Irregular trips without doneChargingTime value: {charging_data_wo_doneChargingTime.shape[0]}")

Regular trips: 60729
Irregular trips without doneChargingTime value: 4081


## 7. Handle User Input

There are three possible scenarios for userInputs:
1. No user inputs
2. Exactly one user input
3. Several user inputs

Possible solutions for handling the userInputs:
1. Leave the userInput attributes in the list column
2. Create one column per userInputs attribute --> Each column contains a list of the input attribute values
3. Create one column per userInputs attribute --> Each column contains only the last value for the input attribute

In [ ]:
print("No user inputs: \n", charging_data[charging_data["userInputs"].isnull()]["userInputs"].iloc[0])

print("Single user input: \n", charging_data["userInputs"].iloc[0])

print("Multiple user inputs: \n", charging_data[charging_data["userID"] == 1117]["userInputs"].iloc[0])

No user inputs: 
 nan
Single user input: 
 [{'WhPerMile': 250, 'kWhRequested': 25.0, 'milesRequested': 100, 'minutesAvailable': 463, 'modifiedAt': 'Thu, 02 Jan 2020 13:09:39 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 20:51:54 GMT', 'userID': 194}]
Multiple user inputs: 
 [{'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:03 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}, {'WhPerMile': 400, 'kWhRequested': 8.0, 'milesRequested': 20, 'minutesAvailable': 65, 'modifiedAt': 'Thu, 02 Jan 2020 14:00:19 GMT', 'paymentRequired': True, 'requestedDeparture': 'Thu, 02 Jan 2020 15:04:58 GMT', 'userID': 1117}]


TBD What we do with this problem: Maybe just create more than one dataset

In [18]:
# Handling of user input

## ... Maybe: Give all entries a increasing integer as ID instead ob the ID as an object

In [19]:
charging_data.sort_index(inplace=True)
charging_data["id"] = range(1, len(charging_data) + 1)
charging_data.head()

id            connectionTime            disconnectTime  \
0   1 2020-01-02 05:08:54-08:00 2020-01-02 11:11:15-08:00   
1   2 2020-01-02 05:36:50-08:00 2020-01-02 14:38:21-08:00   
2   3 2020-01-02 05:56:35-08:00 2020-01-02 16:39:22-08:00   
3   4 2020-01-02 05:59:58-08:00 2020-01-02 08:38:39-08:00   
4   5 2020-01-02 06:00:01-08:00 2020-01-02 14:08:40-08:00   

           doneChargingTime  kWhDelivered  \
0 2020-01-02 09:31:35-08:00        25.016   
1 2020-01-02 12:18:05-08:00        33.097   
2 2020-01-02 08:35:06-08:00         6.521   
3 2020-01-02 07:18:45-08:00         2.355   
4 2020-01-02 10:17:30-08:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  \
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...   
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...   
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...   
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...   
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...   

            next_connection           next_disconnect  
0 2020-01-02 11:30:53-08:00 2020-01-02 15:18:39-08:00  
1 2020-01-02 16:35:27-08:00 2020-01-02 18:42:08-08:00  
2 2020-01-06 05:58:18-08:00 2020-01-06 13:52:37-08:00  
3 2020-01-02 08:48:03-08:00 2020-01-02 14:18:12-08:00  
4 2020-01-02 14:32:14-08:00 2020-01-02 19:36:55-08:00

## TBD:
1. User Input Handling
2. Prepare Wheater Data and Connect the Data to the charging data
3. Create on ore more final datasets and maybe export them